[DataLoader] Typing Doc

ejguan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54773 [DataLoader] Typing Doc**
* #54544 [DataLoader] Typing Enforcement for DataPipe at runtime
* #54066 [DataLoader] Typing Enforcement for DataPipe at construct-time
* #54020 [DataLoader] Typing Enforcement for DataPipe at Compile-time
* #54299 [DataLoader] Implement issubtype for type hints

Differential Revision: [D27364246](https://our.internmc.facebook.com/intern/diff/D27364246)

ejguan committed:
[DataLoader] Typing Doc

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Doc"




[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Doc"


Differential Revision: [D27364246](https://our.internmc.facebook.com/intern/diff/D27364246)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Doc"


Differential Revision: [D27364246](https://our.internmc.facebook.com/intern/diff/D27364246)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Doc"


Differential Revision: [D27364246](https://our.internmc.facebook.com/intern/diff/D27364246)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Doc"


Differential Revision: [D27364246](https://our.internmc.facebook.com/intern/diff/D27364246)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Doc"


Differential Revision: [D27364246](https://our.internmc.facebook.com/intern/diff/D27364246)

[ghstack-poisoned]

VitalyFedyunin reviewed (approved):
Awesome doc, can you please link it from wiki, maybe add DataLoader section into it.

ejguan committed:
Update on "[DataLoader] Typing Doc"


Differential Revision: [D27364246](https://our.internmc.facebook.com/intern/diff/D27364246)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Doc"


Differential Revision: [D27364246](https://our.internmc.facebook.com/intern/diff/D27364246)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Doc"


Differential Revision: [D27364246](https://our.internmc.facebook.com/intern/diff/D27364246)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Doc"


Differential Revision: [D27364246](https://our.internmc.facebook.com/intern/diff/D27364246)

[ghstack-poisoned]

ejguan committed:
Update on "[DataLoader] Typing Doc"


Differential Revision: [D27364246](https://our.internmc.facebook.com/intern/diff/D27364246)

[ghstack-poisoned]
